Add Marstek Venus battery

Chris8er opened the pull request:
This PR adds support for the Marstek Venus battery storage system via Modbus TCP.

**Changes**:
- Adds template templates/definition/meter/marstek-venus.yaml: power measurement (register 32202), state of charge (register 32104), and battery mode control (Normal, Charge, Hold) using registers 42000, 42010, 42020, 43000.
See: https://duravolt.nl/wp-content/uploads/Duravolt-Plug-in-Battery-Modbus.pdf (Duravolt = Marstek Venus)

**Testing**:
- Tested with Marstek Venus firmware version V151 and PUSR USR-DR134 Modbus TCP gateway.
- Verified functionality with real hardware (power, SOC, battery modes).

Chris8er committed:
Add support for Marstek Venus battery via Modbus TCP

Introduce `MarstekVenus` implementation for battery system monitoring and control, including CurrentPower, SoC, and battery mode management. Also, add a template definition for configuration with Modbus TCP parameters.

Chris8er committed:
Set timeout default value to 5s in marstek-venus.yaml

Chris8er committed:
Refactor Marstek Venus control constants and improve clarity

Replaced hardcoded Modbus control values with named constants for better readability and maintainability. Updated associated logic and comments to ensure consistent use of the new constants. Also added `"skiptest"` to the EVCC requirements in the YAML definition.

Chris8er committed:
Use structured logger instead of fmt.Printf in Diagnose.

Chris8er committed:
Simplify battery mode setting by using WriteSingleRegister

Replaced WriteMultipleRegisters calls with WriteSingleRegister to simplify register writing logic. This enhances code readability and reduces unnecessary complexity in setting battery modes. Logging improvements were also added to provide clearer status updates.

Chris8er committed:
Remove duplicate comment in marstek_venus.go

Chris8er committed:
Updated comments to clarify the process of enabling RS485 control for holding the battery in place. Explicitly noted the intentional decision not to disable RS485 to prevent the battery from reverting to Normal mode.

andig commented:
Thank you for this PR. Afaikt by a quick look, there‘s nothing special that would require a Go module here. Could this be implemented purely using templates?

Chris8er commented:
> Thank you for this PR. Afaikt by a quick look, there‘s nothing special that would require a Go module here. Could this be implemented purely using templates?

Ich bin mit den Templates nicht wirklich vertraut. Sehe ich das richtig, dass sie Generatoren für Konfigurationsabschnitte sind?
Kann man darüber für die Steuerung mehrere Modbus Befehle nacheinander senden (über das Modbus Plguin)? Das ist nötig für die aktive Steuerung.

Ich hatte das Ganze schon mal nur über die Config laufen. Allerdings musste ich für die Steuerung dann einen Umweg über IO Broker Script über HTTP Plugin in EVCC gehen.

Chris8er committed:
Remove Marstek Venus Go implementation and replace it with template

Removed the Go-based Marstek Venus meter implementation, replacing it with an enhanced YAML-based custom configuration template. The new template simplifies configuration, adds battery-mode handling, and improves flexibility with support for additional parameters and mappings.

Chris8er committed:
Fix typo in English translation of battery capacity description

Chris8er committed:
Fix scaling factor for total charging energy in Marstek Venus

    Updated the scale value from 1 to 0.01 in the Marstek Venus YAML template. This correction ensures accurate decoding of the total charging energy value from Modbus data.

Chris8er committed:
Enhance parameter descriptions in marstek-venus.yaml

Chris8er committed:
Add configurable parameters for charge power and work mode

Introduced `charge_power` and `work_mode_normal` parameters to allow customization of charge power and work mode settings. Updated Modbus write operations to use these parameters dynamically instead of hardcoded values. Improved flexibility for different deployment scenarios.

andig commented on templates/definition/meter/marstek-venus.yaml:
```diff
@@ -175,15 +187,15 @@ render: |
               decode: uint16
         - source: const
-          value: 2500
+          value: {{ .charge_power }}
```
```suggestion
          value: {{ . maxchargepower }}
```

Chris8er committed:
Update templates/definition/meter/marstek-venus.yaml

Co-authored-by: andig <cpuidle@gmail.com>

Chris8er committed:
Update templates/definition/meter/marstek-venus.yaml

Co-authored-by: andig <cpuidle@gmail.com>

Chris8er committed:
Fix typo in maxchargepower template variable reference

Corrected an erroneous space in the variable reference `{{ . maxchargepower }}` to `{{ .maxchargepower }}`. This ensures proper rendering and functionality in the template.

Chris8er committed:
Remove redundant timeout and delay parameters in YAML files

The `timeout` and `delay` parameters were removed from multiple modbus register definitions in the `marstek-venus.yaml` file. These parameters were either unnecessary or handled elsewhere, simplifying the configuration. This change helps reduce potential configuration errors and improves maintainability.

Chris8er commented:
> Do we need all these timeouts and delays? If not, please remove. Go version didn't have them?

Tatsächlich nicht (getestet). Habe Sie entfernt. Die hatte ich aus einer Vorlage...

Chris8er committed:
Add evcc field and example for Modbus TCP gateway URI

The evcc "skiptest" field was added to indicate test skipping, and an example value for the Modbus TCP gateway URI was provided. These changes improve clarity and usability for configuring the meter definition.

Chris8er committed:
Remove unused "evcc: [\"skiptest\"]" from Marstek Venus config

Example is used for tests, so uri is no longer nil. Tests should pass now.

Chris8er commented:
Tests sollten nun laufen. Musste mich erst ein wenig einlesen.

Chris8er committed:
Remove 'energy' register from marstek-venus template.

The 'energy' register definition was removed because the device does not provide this value via modbus.

andig commented on templates/definition/meter/marstek-venus.yaml:
```diff
@@ -0,0 +1,166 @@
+requirements:
+  description:
+    en: >
+      Requires a Modbus TCP gateway (e.g., PUSR USR-DR134). Tested with firmware version V151.
```
Sorry, discovering new things every time I look. Is that an rs485 to ethernet converter? In this case, does the Marstek have an rs485 connector but no ethernet? Then you should just use the modbus logic:

```yaml
  - name: modbus
    choice: ["rs485"]
    baudrate: 38400
    comset: 8E1
```

andig replied:
Besseres Beispiel: https://github.com/evcc-io/evcc/blob/02d3a362b2d2a515b5ede6c66ab6ab3520f7cc86/templates/definition/charger/daikin-homehub.yaml#L4

andig replied:
Ja und nein. Für Modbus wird der *physikalische* Anschluss konfiguriert, also rs485. Den Rest macht dann das inkludierte Template.

Chris8er committed:
Refactor Marstek Venus template for RS485 Modbus support

Replaced Modbus TCP configuration with RS485-based communication, allowing direct serial or adapter-based connections. Updated parameters, configuration details, and register mappings accordingly. Improved clarity in documentation and streamlined code using templated Modbus configuration.

Chris8er committed:
Add comments to clarify RS485 Control Mode behavior

Comments explain that disabling RS485 Control Mode resets the device and triggers charge/discharge actions. This provides clarity for future modifications and prevents unintended behavior changes.

Chris8er committed:
Fix line break in German translation for meter configuration.

Chris8er committed:
Update description to correct device name from Venus-E to Venus

andig committed:
Apply suggestions from code review

Husches88 commented:
I have a Lilygo T CAN 485 that was set up using a guide from the Tweakers forum. What changes do I need to make to the code or elsewhere so that EVCC can use it?

andig commented:
@Husches88 what does that have to do with Marstek?

andig commented:
If LilyGo is a Modbus converter you can use it. If not... not.

Tunestwo commented:
If I ready correct i first need a modbuss that i can use over wifi? The setup in evcc asks for the ipadress, of the marstek or the modbuss?

DanielGlaas commented:
> If I read it correct i first need a modbuss that i can use over wifi? The setup in evcc asks for the ipadress, of the marstek or the modbuss?

The same question do I also have.
The readme https://docs.evcc.io/docs/devices/meters#marstek-venus-battery-storage does not tell something about an additional device that is needed. This should be improved.

The first stats that a device called "PUSR USR-DR134 Modbus TCP gateway" shall be used. However, this is not a unique name/brand to order.

Chris8er commented:
The evcc documentation specifies that the Marstek Venus battery storage connects via RS485, which can be interfaced using various methods like a TCP Ethernet gateway, TCP WiFi gateway, or a USB-to-RS485 adapter, depending on your setup. The PUSR USR-DR134 Modbus TCP gateway is mentioned as an example (got mine from Amazon), but any compatible RS485-to-TCP gateway should work. However, since evcc supports multiple hardware configurations, I think the docs likely avoid being overly prescriptive to allow flexibility.

DanielGlaas commented:
Well, I agreed that the docs mentions RS485  and that - if you are familiar with this - it is clear that an additional device is needed.

However, all other devices I integrated in my setup worked directly with using their local IP address for the `host` parameter. So the point, that this is not the case for the Marstek Venus was quite surprisingly for me, not having being familiar with the meaning of the term `RS485`.

Chris8er commented:
I'm not part of the evcc team and can't speak for them. Your issue is hardware-specific and beyond evcc's scope, as it only requires RS485 for the Marstek Venus E. The rest is your responsibility. Modbus isn't as straightforward as web APIs, so some technical knowledge is needed, and since it's not plug-and-play, proceed carefully to avoid issues.

Having faced the same challenge connecting the PUSR USR-DR134, I can share a tip. Refer to [this post](https://www.photovoltaikforum.com/thread/247095-marstek-venus-e-hat-jetzt-einen-lan-anschluss-von-mir-bekommen/?postID=4293447#post4293447) for wiring details. I cut the white connector from the Marstek cable and wired it directly to the PUSR. Since Marstek's cable colors may vary across batches, use a multimeter to verify the pin connections before wiring.

DanielGlaas commented:
@Chris8er: Your last reply was very helpful!

I now managed to include my Marstek Venus E into evcc with this device: https://www.amazon.de/dp/B09PD5ZD56/

beebee commented:
> I have an additional question.
> 
> I interprete the code in line 60 (https://github.com/evcc-io/evcc/pull/21487/files#diff-061f0ffca5472bbfa10184a4f4fe0c29ce3bf98874a73f77c1e3a8fe25fd96d5R60) such that evcc is able to control the charging value of the Marstek battery. A variable names `batterymode` is mentioned, but I don't know what to do to enable/use it. `batterymode` appears in the docs only at one place, which is not helpful for me: https://docs.evcc.io/docs/devices/meters#manuell

the batterymode definitions enable how evcc can use the active battery control on the device: https://docs.evcc.io/en/docs/features/battery#active-battery-control

For some you might need to enable experimental mode in the UI config. So it should be possible to grid load the device on cheap prices with by evcc via UI menu for HomeBattery/Grid Charging Tab also.  (batterymode 3 / "Charge")

DanielGlaas commented:
> > I have an additional question.
> > I interprete the code in line 60 (https://github.com/evcc-io/evcc/pull/21487/files#diff-061f0ffca5472bbfa10184a4f4fe0c29ce3bf98874a73f77c1e3a8fe25fd96d5R60) such that evcc is able to control the charging value of the Marstek battery. A variable names `batterymode` is mentioned, but I don't know what to do to enable/use it. `batterymode` appears in the docs only at one place, which is not helpful for me: https://docs.evcc.io/docs/devices/meters#manuell
> 
> the batterymode definitions enable how evcc can use the active battery control on the device: https://docs.evcc.io/en/docs/features/battery#active-battery-control
> 
> For some you might need to enable experimental mode in the UI config. So it should be possible to grid load the device on cheap prices with by evcc via UI menu for HomeBattery/Grid Charging Tab also. (batterymode 3 / "Charge")

@beebee Thanks for the link to the docs. But it does not explain the three modes of `batterymode`.

To be honest, it still is not clear to me what I have to do.

DanielGlaas commented:
> @DanielGlaas : How did you setup the Elfin Device?
> 
> When implementing the Code in the evcc.yaml with the dedicated adresses it computes error message as it seems, that it cannot access the tcp-server:
> 
> ```
> 21:12:17.703778489 +0200 CEST m=+1481.251765055 !! 3013ms read failed: read tcp 192.168.188.80:40902->192.168.188.114:502: i/o timeout
> ```
> 
> in the evcc.yaml:
> 
> * name: my_battery
>   type: template
>   template: marstek-venus
>   usage: battery
>   modbus: rs485tcpip
>   id: 1
>   host: 192.168.188.114 # Hostname
>   port: 502 # Port
> 
> in general the communication is working with modbus in homeassistant... :-?

@tjahebro1
Is homeassitant also communicating via the Elfin device?
If yes, I only see a possible error in the evcc.yaml device and there only in the parameters host and port.

Maybe you compare these parameters with those in homeassistant.

Chris8er commented:
> > Clicking "Battery" in the EVCC menu allows you to configure your battery to charge from the grid (scheduled). This disables discharging and charges the battery at the specified power level (config). The code is used for that.
> > The "work_mode_normal" option in the Venus EVCC settings determines the mode that Venus returns to after being locked by EVCC vehicle fast charge.
> 
> @Chris8er Thank you for the additional details.
> 
> To sum up, the setup of the Marstek with a shelly meter is necessary, such that it can autonomously decide without evcc when to charge/discharge. Charging/Discharging can not be completely controlled by evcc, which was my initial expectation/hope.

Yes, you need a meter compatible with Venus to effectively manage battery adaptation to your grid consumption. The cycles in EVCC might be too slow for this purpose anyways. Have you installed a meter? If so, you could potentially use it with [Unimeter](https://github.com/sdeigm/uni-meter).

tjahebro1 commented:
> > @DanielGlaas : How did you setup the Elfin Device?
> > When implementing the Code in the evcc.yaml with the dedicated adresses it computes error message as it seems, that it cannot access the tcp-server:
> > ```
> > 21:12:17.703778489 +0200 CEST m=+1481.251765055 !! 3013ms read failed: read tcp 192.168.188.80:40902->192.168.188.114:502: i/o timeout
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in the evcc.yaml:
> > 
> > * name: my_battery
> >   type: template
> >   template: marstek-venus
> >   usage: battery
> >   modbus: rs485tcpip
> >   id: 1
> >   host: 192.168.188.114 # Hostname
> >   port: 502 # Port
> > 
> > in general the communication is working with modbus in homeassistant... :-?
> 
> @tjahebro1 Is homeassitant also communicating via the Elfin device? If yes, I only see a possible error in the evcc.yaml device and there only in the parameters host and port.
> 
> Maybe you compare these parameters with those in homeassistant.


solved: 
  - name: my_battery
    type: template
    template: marstek-venus
    usage: battery
    # RS485 via TCP/IP (Modbus RTU)
    **modbus: tcpip**
    id: 1
    host: 192.168.188.114 # Hostname
    port: 502 # Port

it did not work with "rs485tcpip" in my config using the elfin modbus to wifi adapter. changed now to "tcpip". now it seems to work.

Tunestwo commented:
> This should be an configuration (assignment) issue. Nothing to du with the Venus integration itself.

Hmm okey… going to look for a fix, not happend to anyone before?

Chris8er commented:
EVCC does not directly support dynamic grid consumption control, as it primarily integrates batteries to adjust EV charging power. As mentioned, EVCC's cycle concept is too slow for real-time grid management. It mainly handles the scenarios "stop charging during fast EV charging," "prevent battery discharge while charging an EV," and "charge at full power when prices are low." For broader control, use Venus with a compatible meter.

njoerd114 commented:
For anybody who's using the ha integration through lilygo - you can integrate with the HA REST API like this:

```yaml
power: # current power
  source: http
  uri: http://YOURHOMEASSISTANT/api/states/sensor.esp_akku_marstek_ac_power
  auth:
    type: bearer
    token: YOURLONGLIVEDACCESSTOKEN
  jq: '.state | tostring'

soc: # state of charge in % (for battery)
  source: http
  uri: http://YOURHOMEASSISTANT/api/states/sensor.esp_akku_marstek_battery_state_of_charge
  auth:
    type: bearer
    token: YOURLONGLIVEDACCESSTOKEN
  jq: '.state | tostring'
```

I'm going to add controls later...

joernbeutel commented:
Maybe someone knows the answer: I have tried to incorporate the Venus connected to **Lilygo**  to evcc. It doesn't work for me. According to tweakers, however, it works. Anybody who succeded can push me round the corner? I have tried with yaml as well as with the GUI.
![Screenshot 2025-06-13 120613](https://github.com/user-attachments/assets/7e66dc5d-31a5-4eec-9200-abfcb0d50c1a)
![Screenshot 2025-06-13 120512](https://github.com/user-attachments/assets/ec43dca1-7f2a-41dc-bd9c-fca5a50d8a18)

joernbeutel commented:
@njoerd114 Ah, THANK YOU!! okay. Either - or. I have it in homeassistant using esphome. Now it is clear that some got it working. Prob. not using esphome...
As you probably already guessed by me not having a clue of what you indicated in your example I guess I have to find out how to create a rest API in evcc, right? Then I could use yor script snipes...?

joernbeutel commented:
I treid to find out using chatgpt. Was advised to put it under meters like this
`

- name: Venus1
  type: custom
  power:
    source: http
    uri: http://1xxxxx111:8123/api/states/sensor.esp_akku_marstek_ac_power
    auth:
      type: bearer`...........

but obviously is wrong.

Alanin commented:
> @Chris8er: Your last reply was very helpful!
> 
> I now managed to include my Marstek Venus E into evcc with this device: https://www.amazon.de/dp/B09PD5ZD56/

May I ask how you connected the wires? I have 5 coming out of the marstek and only 4 pins in the adapter...

DanielGlaas commented:
> > @Chris8er: Your last reply was very helpful!
> > I now managed to include my Marstek Venus E into evcc with this device: https://www.amazon.de/dp/B09PD5ZD56/
> 
> May I ask how you connected the wires? I have 5 coming out of the marstek and only 4 pins in the adapter...

@Alanin: Sorry for my late reply, I had some family issues.

I attached a picture of my wiring:
![IMG_20250720_161026](https://github.com/user-attachments/assets/656afcd5-37b1-4182-b23b-c9174aec0e79)

The explanation:
1. The red and yellow wires at both sides are drilled inside Marstek cable, which shows that these are for the data. However, which one is yellow is red, that I had to guess.
2. The two black one are +5V and GND, that is static and you can measure it with a voltmeter.
3. The red wire in the air seems not be needed, I didn't find a usage for it. The setup works since several weeks without this wire being connected.

Alanin commented:
Thank you a lot! I have tried with jumper cables. Turned out that the adapter is not working...
Returned to Amazon. 🫣

JonasD1986 commented:
Hi,

I’m using a Marstek Venus E 3.0 battery together with a CT003 P1 meter.

The battery works perfectly via the Marstek app (charging/discharging, SOC visible).

Hardware interfaces available: Wi-Fi, LAN, and RS485.

Over Wi-Fi, the battery receives an IP (e.g., 192.168.0.xxx).

Ping works fine.

But Modbus TCP port 502 is refused → EVCC cannot connect.


The LAN port does not show up in my router/DHCP → seems inactive.

RS485 is available, but I’d like to avoid extra cabling/adapters if possible.


My questions:

Is the Marstek Venus E 3.0 officially supported via Modbus TCP over Wi-Fi or LAN in EVCC?

Does Modbus TCP need to be explicitly enabled (installer mode / hidden menu / firmware)?

Or is RS485 currently the only way to integrate the Venus into EVCC and Home Assistant?


Context:

EVCC version: v0.207.6

Venus firmware version: v114

Models: Venus E 3.0 + CT003 P1


Any tips or working configs would be highly appreciated. I’m also happy to test further if needed.

Thanks!

ebbenberg commented:
@JonasD1986 
There is still no official modbus register document for the V3.
Some guy (dutch tweakers forum) tried to readout the modbus RTU with the v2 registers, but a lot of registers missing or are different.
I even wonder if modbus tcp/ip will be available
At the moment (i think) things are still under development for the V3 and not ready. 
But can be fixed with firmware update from marstek.

This is for me a reason not to buy the v3 at the moment (i like to have a second battery)

pergolafabio commented:
Hi @Chris8er , there is now also en local API using UDP to use with Venus, are you willing to add that support too? Then no additional hardware is needed anymore...

https://eu.hamedata.com/ems/resource/agreement/MarstekDeviceOpenApi.pdf

To enable the API, you need to send a feedback with app

pergolafabio commented:
Yeah, I know, but I have the "advanced features" in the app, so I can enable it there too.. would be great if we can use the API in EVCC, no additional hardware is needed then
